lib/db/mysql/scripts/purge-expired-tokens.sql

```diff
@@ -0,0 +1,43 @@
+DROP PROCEDURE IF EXISTS PURGE_EXPIRED_TOKENS;
+
+DELIMITER $$
+# Create a procedure that will delete a total of `roundCount` * `limitDeletePerRound` items, every `delaySeconds`.
+# Skip tokens that belong clientId
+CREATE PROCEDURE PURGE_EXPIRED_TOKENS(IN roundCount INT, IN limitDeletePerRound INT, IN delaySeconds INT, IN ignoreClientId CHAR(10))
+
+BEGIN
+  DECLARE startIndex INT DEFAULT 1 ;
+  DECLARE endCount INT DEFAULT roundCount ;
+  DECLARE expiredItemCount INT DEFAULT 0;
+
+  PURGE_LOOP: LOOP
+
+	# Exit loop if no expired items found
+    SET expiredItemCount = (SELECT COUNT(*) FROM tokens WHERE expiresAt < NOW() AND UNHEX(clientId) != ignoreClientId ORDER BY expiresAt);
+    IF expiredItemCount = 0 THEN
+		LEAVE PURGE_LOOP;
+	END IF;
+
+	# Perform a round of expired token deletions
+	SET SQL_SAFE_UPDATES = 0;
+	DELETE p1 FROM `fxa_oauth`.`tokens` AS p1
+	INNER JOIN (
+		SELECT token FROM `fxa_oauth`.`tokens` WHERE expiresAt < NOW() AND UNHEX(clientId) != ignoreClientId ORDER BY expiresAt ASC LIMIT limitDeletePerRound
+	) AS P2
+	USING (token);
```

Comment:
In order to do this efficiently, we'll need to add an index on the `expiresAt` column.  Without it MySQL will have to do a full table scan on every iteration of the loop:

```
mysql> EXPLAIN DELETE p1 FROM `fxa_oauth`.`tokens` AS p1 INNER JOIN (SELECT token FROM `fxa_oauth`.`tokens` WHERE expiresAt < NOW() AND UNHEX(clientId) != '123445' ORDER BY expiresAt ASC LIMIT 100) AS P2;
+----+-------------+------------+------+---------------+------+---------+------+-------+---------------------------------------+
| id | select_type | table      | type | possible_keys | key  | key_len | ref  | rows  | Extra                                 |
+----+-------------+------------+------+---------------+------+---------+------+-------+---------------------------------------+
|  1 | PRIMARY     | <derived2> | ALL  | NULL          | NULL | NULL    | NULL |   100 | NULL                                  |
|  1 | PRIMARY     | p1         | ALL  | NULL          | NULL | NULL    | NULL | 34077 | Using join buffer (Block Nested Loop) |
|  2 | DERIVED     | tokens     | ALL  | NULL          | NULL | NULL    | NULL | 34077 | Using where; Using filesort           |
+----+-------------+------------+------+---------------+------+---------+------+-------+---------------------------------------+
3 rows in set (0.17 sec)
```

The `Using where; Using filesort` here is basically MySQL telling us it's going to have a bad time :-)

The `DELETE LIMIT` form is better as it can avoid sorting all the rows, but `type=ALL` tells us it'll still need a full table scan:

```
mysql> EXPLAIN DELETE  FROM `fxa_oauth`.`tokens` WHERE expiresAt < NOW() AND UNHEX(clientId) != '123445' LIMIT 100;
+----+-------------+--------+------+---------------+------+---------+------+-------+-------------+
| id | select_type | table  | type | possible_keys | key  | key_len | ref  | rows  | Extra       |
+----+-------------+--------+------+---------------+------+---------+------+-------+-------------+
|  1 | SIMPLE      | tokens | ALL  | NULL          | NULL | NULL    | NULL | 34077 | Using where |
+----+-------------+--------+------+---------------+------+---------+------+-------+-------------+
1 row in set (0.00 sec)
```

I can't think of a good way to do this efficiently without adding an index on `expiresAt`.